### ISSUE TITLE
Help Wanted - Turn into true cross-platform

### ISSUE THREAD

**Issue by FrankPfattheicher** [chromelyapps/Chromely#66]
I want to help here: 

> Turn CefGlue implementation into true cross-platform. 
> This will merge Chromely.CefGlue.Winapi, Chromely.CefGlue.Gtk, Chromely.CefGlue.Mac into Chromely.CefGlue. 

The first problem would be that "CefRuntime" is defined in more than one assembly...
I would prefere to introduce an interface here.

How can we discuss the way to go in advance?
This issue tracker would be not the best place?

**Comment by mattkol** [comment]
_"I want to help here:"_
That will be great ..

I am also looking into migrating Chromely to an organization github project. Just have not gotten around to doing it, but if you will be actively involved now, I can make it faster ...

Back to the question .. we can use the [![Join the chat at https://gitter.im/chromely_/Lobby](https://badges.gitter.im/chromely_/Lobby.svg)](https://gitter.im/chromely_/Lobby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) or my email.

Just an update .. I started looking into the Cross platform task during Christmas. I started POCs in part for Mac, just really at rudimentary stages for now .. I am looking at using [Libui](https://github.com/andlabs/libui) for the thin host and of course CefGlue. 

My research on CefGlue so far is there is no official support or sample for Mac, but looks like all the needed code is there. So this is the main risk, but I am willing to take a stab at it and see what happens.

You mentioned the CefRuntime being defined in different assembly .. I welcome your approach to dealing with that. In my own investigation though the biggest issue is with Dllimports. I think when CefGlue was started all the CEF primary dll are named libcef (libcef.dll, libcef.so, libcef.dylib). Currently for Mac libcef.dylib is renamed to "Chromium Embedded Framework.framework" by [CEF](https://stackoverflow.com/questions/47538368/how-to-properly-map-an-expected-dll-name-to-a-native-mac-framework-for-mono?rq=1). The challenge here is to have those libraries loaded dynamically. The alternative is to have separate platform dll which makes it less "truly cross platform"

**Comment by mattkol** [comment]
@FrankPfattheicher Chromely is now turned to an organization - chromelyapps. Please let me know if should invite you. Thanks.

**Comment by FrankPfattheicher** [comment]
@mattkol Great ! - Yes please invite me, i have planned to start next week with the code reorganisation to be "true cross-platform".

**Comment by mattkol** [comment]
Great welcome aboard! Please let me know if I got the privileges right. First time I am doing this. Thanks.

**Comment by mattkol** [comment]
@FrankPfattheicher your membership visibility is currently "Private". If that is what you prefer that is ok .. if it needs to be changed, it looks like only you can change it from your Profile Settings. Thanks.

**Comment by FrankPfattheicher** [comment]
@mattkol i changed it to public.

**Comment by FrankPfattheicher** [comment]
Wanted to be shure not destroying something i tried to run the demo samples.
I now ars somewhat frustrated.
* Chromely.CefGlue.Winapi.Demo (No error but empty window)
* Chromely.CefGlue.Winapi.netCoreDemo (The program '[9456] dotnet.exe' has exited with code -2147450751 (0x80008081).)
* Chromely.CefGlue.Gtk.Linux.Demo (see other issue)
Got not one of that running :-(

**Comment by FrankPfattheicher** [comment]
I found "build\\win\chromelycef.exe" download v68 --cef-binary-version=3.3440.1805.gbe070f9 --cpu=AnyCPU --dest="\bin\Debug\netcoreapp2.1\" is loading a wrong version for my core sample.

I have now written an other loader to be integrated in the resulting application.
So there is no build step necessary. The resulting exe can be published without the cef binaries.

I used the CefRuntime.Load() to check if the matching binaries are installed.
If not i download the propper binaries. Interesting?

**Comment by mattkol** [comment]
_I used the CefRuntime.Load() to check if the matching binaries are installed.
If not i download the propper binaries. Interesting?_

That is really great, I like that approach.

**Comment by mattkol** [comment]
_Chromely.CefGlue.Winapi.Demo (No error but empty window)_

This is likely correct, If you build the demo, using the project reference instead of the nuget package reference, it will require downloading the CEF binary manually.

So in project file instead of:
````csharp
  <ProjectReference Include="..\..\Chromely.CefGlue.Winapi\Chromely.CefGlue.Winapi.csproj" />
````
It will be:
````csharp
  <ProjectReference to nuget package />
````

**Comment by FrankPfattheicher** [comment]
Hi @mattkol 
after a while of doing nothing to the project i now want to do more.

Have still problems running it :-(
Created a sample with my bachend library: https://github.com/FrankPfattheicher/stonehenge.Chromely
Included my own runtime loader here - to be moved to the project later...

After the application starts it terminates a short time later !???
[0207/133117.870:ERROR:gl_surface_egl.cc(863)] eglInitialize D3D11 failed with error EGL_NOT_INITIALIZED, trying next display type
OpenGL Warning: crPixelCopy3D:  simply crMemcpy'ing from srcPtr to dstPtr

Any ideas?

**Comment by FrankPfattheicher** [comment]
Further investigations...
I found "CefRuntime.ExecuteProcess() returns -1"
After this it starts the application process two more times !???
Shouldn't it terminate on that error?

**Comment by mattkol** [comment]
_[0207/133117.870:ERROR:gl_surface_egl.cc(863)] eglInitialize D3D11 failed with error EGL_NOT_INITIALIZED, trying next display type
OpenGL Warning: crPixelCopy3D: simply crMemcpy'ing from srcPtr to dstPtr_

@FrankPfattheicher not an error that I am familiar with. Is this is on Windows or Linux. If Windows I can look ASAP. If Linux, I have to setup my VM all over again.

**Comment by mattkol** [comment]
_"CefRuntime.ExecuteProcess() returns -1"_

@FrankPfattheicher I have seen the situation since the project started and cannot find an explanation on why it does not terminate on error. It looks like it ignores it for subprocess launch.

The reason it runs 3 times, is because the sub-processes are also launched. Please check #69 for more info.

**Comment by mattkol** [comment]
@FrankPfattheicher  I am able to run the Windows version. It comes on and closes with an error in Winapi. Looks like it has to do with the Kestrel host. 

![image](https://user-images.githubusercontent.com/18384207/52519363-c4ddf680-2c1f-11e9-8d70-1f787c29df6a.png)

Error:
FailFast:
A callback was made on a garbage collected delegate of type 'WinApi!WinApi.User32.WindowProc::Invoke'.

   at WinApi.User32.User32Methods.GetMessage(WinApi.User32.Message ByRef, IntPtr, UInt32, UInt32)
   at WinApi.User32.User32Methods.GetMessage(WinApi.User32.Message ByRef, IntPtr, UInt32, UInt32)
   at Chromely.CefGlue.Winapi.BrowserWindow.NativeWindow.RunMessageLoop()
   at Chromely.CefGlue.Winapi.BrowserWindow.HostBase.RunInternal(System.String[])
   at Chromely.CefGlue.Winapi.BrowserWindow.HostBase.Run(System.String[])
   at StonehengeChromelySample.Program.Main(System.String[])

Looks like Chromely is being passed a live URL. Is that the intention? Chromely is not expected to work that way out of the box. If external url is expected it should be [registered](https://github.com/chromelyapps/Chromely/wiki/External-Url-Registration). Or a [custom scheme](https://github.com/chromelyapps/Chromely/wiki/Custom-Scheme-Handling) should be created.

Another alternative is to start Chromely with non-live html page and launch Kestrel url the way "HTML5 Tests", "HTML6 Tests" and "Google Shaka Support Tests" tests are launched in the demos. 
https://github.com/chromelyapps/Chromely/blob/807cc53d02783b4fcdb3d07b7c6a55da50e461df/src/Demos/Chromely.CefGlue.Winapi.Demo/app/chromely.html#L115

**Comment by mattkol** [comment]
@FrankPfattheicher in Windows, I am able to figure out what could be the reason. Looks like the KestrelHost and Chromely are running in the same process and causing collision. It will be easier to run KestrelHost in a separate process. I created a console app for KestrelHost and I am able to run successfully. If you don't want to create separate console opp then you can try AppDomain or Start a process.

**Comment by FrankPfattheicher** [comment]
@mattkol I now got everythin running :-)
On Linux an additional commandline switch no-zygote made the thing!
On Windows it runs net framework 4.7.2 and netcoreapp 2.1.
Thanks for the hints.

On Linux (Ubuntu 18.10) the window resizing does not work - but not relevant yet.

**Comment by FrankPfattheicher** [comment]
@mattkol on Linux i use GitKraken as git client.
I am currently not able to push something - you should got an mail to enable this !?
I now can start "Turn into true cross-platform".

I first will port my CefLoader to a new assembly within the project.
How do you build and publish the nuget packages? It should be there...

**Comment by mattkol** [comment]
@FrankPfattheicher on Linux resize this issue has been there from day one. I could not figure out. Please see [Not Working Yet](https://github.com/chromelyapps/Chromely/wiki/Getting-Started-CefGlue-Gtk-(Linux)#not-working-yet)  and https://github.com/chromelyapps/Chromely/issues/48#issuecomment-444745031

**Comment by mattkol** [comment]
@FrankPfattheicher thanks I tried the CefLoader, and it looks simpler than what was created earlier. 

I have a comment though .. when Chromely started I wanted to have as little dependencies as possible. This is why the Simple logger, container and JsonMapper are embedded. These are all smaller files. And users can create custom ones if they do not like them. I decided on embedding these code because I want  developers to have less headaches worrying about dependencies.

I see that you use external SharpZipLib as external nuget package. What I did is to extract the files needed from SharpZipLib so that there won't be need for dependencies. Nothing is written in stone, if the external nuget is preferred by you, I am fine with that too.

This is the folder for SharpZipLib - https://github.com/chromelyapps/cef-binaries-downloader/tree/master/src/CefDownloaderNetCore/SharpZipLib

Thanks.

**Comment by mattkol** [comment]
Hi Frank,
Just to let you know the project has Resharper support.
If you are interested, I think I should be able to add you to the license.
Thanks,
mattkol

**Comment by FrankPfattheicher** [comment]
Hi @mattkol, i'm also a friend of "as little dependencies as possible".
Due SharpZipLib is also MIT licensed i will use a local copy also when integrating the loader into chromelyapp.
I still look for a "good" method informing the user about the download - it tooks a long time...

Another question:
Where can i find that CEF loading at build time ???
On my machine it always loads a bad version resulting in another download at program start :-(

**Comment by FrankPfattheicher** [comment]
Resharper
I have a full featured JetBrains license for my own - it's worth every cent ;-)

**Comment by mattkol** [comment]
To load the right format, the project CPU must be set (x64, x86). If not set it defaults to x86. I should have forced to x64, Also if you are running .NET Core the version is not properly selected because .NET and .NET Core builds are different. Just needs overhauling.

The workflow is:
1. It checks the bin first .. if not there it checks the nuget folder.
2. If nothing is cached in the nuget folder, it does the download.
3. If binaries are in nuget folder it copies them to the bin.

**I just think it is broken somewhere ...**

The downloader chromelycef.exe chromelycef.dll is in the nuget folder.
![image](https://user-images.githubusercontent.com/18384207/52526246-59297700-2c7b-11e9-8b7a-155d088c467d.png)

So if we use the new approach, we have to delete it in the project setting..
1. We delete this folder - https://github.com/chromelyapps/Chromely/tree/master/src/Chromely.CefGlue.Winapi/build
2. Remove the reference from the project file -https://github.com/chromelyapps/Chromely/blob/f2143aead2408dee19bf4986b9f9acb78bf3c328/src/Chromely.CefGlue.Winapi/Chromely.CefGlue.Winapi.csproj#L56

This can be done in CefGlue v70 or v68.0.0.1.

**Comment by mattkol** [comment]
I think I found out what the issue is. 
This file - widevinecdmadapter.dll that was listed as a requirement seems no longer included and since it cannot find it, it is reloading the entire files.

https://github.com/chromelyapps/cef-binaries-downloader/blob/84e7bc48957a704794fc905e8159e8b4df5f0f76/src/CefDownloader/DownloadService.cs#L219

**Comment by mattkol** [comment]
I copied the file into the bin folder. This will stop the re-loading.

[widevinecdmadapter.txt](https://github.com/chromelyapps/Chromely/files/2848304/widevinecdmadapter.txt) Please rename the file to widevinecdmadapter.dll.

**Comment by FrankPfattheicher** [comment]
@mattkol last week i tried to fix that Linux resize problem - many hours - no success.
So i returned to my sample trying to reduce all code to only ONE base. Done.
Next step will be to create an new assemblies within Chromely named 
"Chromely.CefGlue" which will contain all common CefGlue code and
"Chromely.CefGlue.Loader" with my version of the loader.
I already removed the SharpZipLib dependency as said.
Any comments?

**Comment by FrankPfattheicher** [comment]
@mattkol i don't know how the nuget packages are build and published.

For my own packages i am using Azure DevOps for automated build - test - pack - publish
triggered on github checkins.

How is it done here?

**Comment by mattkol** [comment]
@FrankPfattheicher I am okay with different assemblies or folder links. Whichever is easier to manage. We can also embed them as resources. If I will make a choice I will start with folder link. Especially for Chromely.CefGlue. For the downloader, if you think it will be useful for others like other CEF projects then a nuget for external use and folder link for internal use may be great.  Just my choice. But either way is great, as long as it is easy to manage and developers would not give us more work to do on missing references. 

On the publishing of nugets, I just logged in with my Microsoft account on Nuget.org and upload. The Azure DevOps path is the ultimate, I started it but had funny issues and looks like I lost the link. Please see the PR. I spent a whole Saturday on that weeks ago - it is a lot of work maintaining a project like this, thank God you are on board.

**Comment by mattkol** [comment]
@FrankPfattheicher also I usually publish the release versions. You don't have to set anything, the packages will be built when you build the projects.

**Comment by FrankPfattheicher** [comment]
@mattkol maybe the best will be to create a free Azure DevOps account for "chromelyapps".

I did it immediately: https://dev.azure.com/fpf0438/Chromely%20Apps

You should got an invitation mail.
What do you think to use the backlog there?
The issues are better on github but resulting tasks can be added to DevOps.

I can setup the pipeline later...

**Comment by mattkol** [comment]
@FrankPfattheicher that is great! That will help with collaboration. Can we add @Fiahblade too to the Azure DevOps invite too.

I may not be able to look closely at anything until weekend.

Huge progress! Thanks.

**Comment by FrankPfattheicher** [comment]
@Fiahblade i need an email address to invite you...

**Comment by FrankPfattheicher** [comment]
Azure DevOps renamed to https://dev.azure.com/ChromelyApps/Chromely

**Comment by FrankPfattheicher** [comment]
@mattkol
"A request to install Azure Pipelines has been submitted on the @chromelyapps account."
Can you please accept the request...

**Comment by mattkol** [comment]
@FrankPfattheicher done!
Thanks.

**Comment by FrankPfattheicher** [comment]
@mattkol  Does nor work :-(  (Not an Administrator)
Can you please try to add the github connection:
https://dev.azure.com/ChromelyApps/Chromely/_settings/boards-external-integration

**Comment by FrankPfattheicher** [comment]
Azure DevOps needs to add webhooks to enable CI automated builds on check-in...

**Comment by FrankPfattheicher** [comment]
@mattkol 
Unable to configure a service on the selected GitHub repository. This is likely caused by not having the necessary permission to manage hooks for the selected repository.

Damn shit: Please try to enable
"Enable continuous integration" in the "Chromely-CI" build pipeline

**Comment by mattkol** [comment]
@FrankPfattheicher done!
Thanks.

**Comment by FrankPfattheicher** [comment]
@mattkol  I think we can close this issue here.
Next steps on Azure DevOps platform...

**Comment by FrankPfattheicher** [comment]
@mattkol i now started with the cross-platform branch...
If you like please add more ideas in the backlog on
https://dev.azure.com/ChromelyApps/Chromely/_backlogs/backlog/Chromely%20Team/Stories

**Comment by mattkol** [comment]
@FrankPfattheicher certainly will do. 
Thanks.

**Comment by jwosty** [comment]
I have experience working with MacOS APIs with C# (Xamarin.Mac), and I want to help make this project work on MacOS. Where would be the best place to look to start? Is there some sort of overview of the internal structure?

**Comment by mattkol** [comment]
@jwosty that will be great!

There is really no Architecture diagram for Chromely to explain internal structure, but there should be. We can work on that.

But basically the principle here is to host Chromium via CefSharp, CefGlue implementation in a "thin" native hosts.

By "thin" native hosts, we mean no child controls like "TextBox", "Label", "Button", "DataGrid" etc of the native UI involved - just the main "window widget", if you may. All child controls are going to be HTML based and left to the developers. This is why WinForms and WPF are not supported.

The most recent approach for true-cross platform is is using [LibUI](https://github.com/andlabs/libui) - code here - [LibUI Branch](https://github.com/chromelyapps/Chromely/tree/cross-platform-libui). This approach works on Mac - you can try the binaries here - [chromely_mac.zip - entry exe file - TestCore](https://github.com/chromelyapps/Chromely/files/3276265/chromely_mac.zip). Interestingly Windows and Mac work, but Linux is not working yet. I personally have not had the bandwidth to look closely for sometime, and not sure if other team members have looked too.

Having said that, we do not have to use LibUI we can use Xamarin.Mac too and probably make LibUI an option when we find a fix.

Each time I look at Xamarin.Mac, I just find it difficult to extract the "window" only. But with your expertise we can. My take is we should avoid the child controls so that we keep Chromely as simple as possible.

-----------------------
To understand Chromely, I will suggest you take it for a ride using the [Demos](https://github.com/chromelyapps/demo-projects/tree/master/regular-chromely), checkout the [Wiki](https://github.com/chromelyapps/Chromely/wiki) and shoot us @FrankPfattheicher @Fiahblade questions or we can create an issue to track this. If the architecture is needed I can spend sometime on that. 

Projects structure:

       - CefSharp - Windows
           - Chromely.Core
           - Chromely.CefSharp.Winapi

       - CefGlue - Windows
           - Chromely.Core
           - Chromely.CefGlue
           - Chromely.CefGlue.Winapi

       - CefGlue - Linux
           - Chromely.Core
           - Chromely.CefGlue
           - Chromely.CefGlue.Gtk

So if we are going to use Xamarin.Mac we can have:

       - CefGlue - Mac
           - Chromely.Core
           - Chromely.CefGlue
           - Chromely.CefGlue.Xamarin.Mac (or Chromely.CefGlue.Mac or Chromely.CefGlue.Xamarin) 


Every implementation requires a core/common lib - Chromely.Core. 
CefGlue (that is need for cross-platform) also has a common lib - Chromely.CefGlue. Then a more specific lib - Chromely.CefSharp.Winapi
Chromely.CefGlue.Winapi, Chromely.CefGlue.Gtk.

So the specific lib is in format:
Chromely.[CefGlue|CefSharp].NativeHostType

**Comment by mika76** [comment]
Hey guys - is there maybe an open macos issue/PR we could follow to get notifications of progress? I ask because this issue is closed...

**Comment by mattkol** [comment]
@mika76 yes we could do that with an issue tracker. 

But there is a dedicated branch -
--  using [LibUI](https://github.com/andlabs/libui) - code here - [LibUI Branch](https://github.com/chromelyapps/Chromely/tree/cross-platform-libui). 
----- This approach works on Mac - you can try the binaries here - [chromely_mac.zip - entry exe file - TestCore](https://github.com/chromelyapps/Chromely/files/3276265/chromely_mac.zip). Interestingly Windows and Mac work, but Linux is not working yet. 

The bottom line is we need help with MacOS. Someone who understands MacOS well could take us to the next level on this.
